casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -122,20 +126,44 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Log: Time: SafetyOracle: Las
   def estimator(dag: BlockDagRepresentation[F]): F[IndexedSeq[BlockHash]] =
     Estimator[F].tips(dag, approvedBlock).map(_.tips)
 
-  def lastFinalizedBlock: F[BlockMessage] =
-    for {
-      dag                    <- blockDag
-      lastFinalizedBlockHash <- LastFinalizedStorage[F].getOrElse(approvedBlock.blockHash)
-      updatedLastFinalizedBlockHash <- LastFinalizedBlockCalculator[F]
-                                        .run(dag, lastFinalizedBlockHash)
-      _ <- LastFinalizedStorage[F].put(updatedLastFinalizedBlockHash)
-      _ <- EventPublisher[F]
-            .publish(
-              RChainEvent.blockFinalised(updatedLastFinalizedBlockHash.base16String)
+  def lastFinalizedBlock: F[BlockMessage] = {
+
+    def finalisationEffect(finalizedChildHash: BlockHash): F[Unit] =
+      for {
+        block          <- BlockStore[F].getUnsafe(finalizedChildHash)
+        deploys        = block.body.deploys.map(_.deploy)
+        deploysRemoved <- DeployStorage[F].remove(deploys)
+        _ <- Log[F].info(
+              s"Removed $deploysRemoved deploys from deploy history as we finalized block ${PrettyPrinter
+                .buildString(finalizedChildHash)}."
             )
-            .whenA(lastFinalizedBlockHash != updatedLastFinalizedBlockHash)
-      blockMessage <- BlockStore[F].getUnsafe(updatedLastFinalizedBlockHash)
+        _ <- BlockDagStorage[F].addFinalizedBlockHash(finalizedChildHash)
```

Comment:
Where is indirectly finalized block persisted? This is just in memory update.

Reply:
OK, so this means indirectly finalized blocks will also have finalized state after restart?

Reply:
Yes, as last finalized block is persisted